n3fit/src/evolven3fit_new/evolve.py

```diff
@@ -0,0 +1,221 @@
+
+from ekobox import genpdf, gen_info
+from ekomark import apply
+from eko import basis_rotation as br
+from eko import output
+from validphys.api import API
+
+from . import utils, eko_utils
+
+
+log = logging.getLogger(__name__)
+
+LOG_FILE = "evolven3fit_new.log"
+
+LOGGING_SETTINGS = {"formatter" : "%(asctime)s %(name)s/%(levelname)s: %(message)s",
+                    "level" : logging.INFO
+}
+
+def evolve_fit(
+    conf_folder,
+    q_fin,
+    q_points,
+    op_card_dict,
+    t_card_dict,
+    force,
+    eko_path=None,
+    dump_eko=None,
+):
+    """
+    Evolves all the fitted replica in conf_folder/nnfit
+
+    Parameters
+    ----------
+
+        conf_folder: str or pathlib.Path
+            path to the folder containing the fit
+        q_fin: float
+            final point of the q_grid
+        q_points: int
+            number of points in the q_grid
+        op_card_dict: dict
+            user settings for the op_card
+        t_card_dict: dict
+            user settings for the t_card
+        force: bool
+            whether to force the evolution to be done again
+        eko_path: str or pathlib.Path
+            path where the eko is stored (if None the eko will be
+            recomputed)
+        dump_eko: str or pathlib.Path
+            path where the eko is dumped (if None the eko won't be
+            stored)
+    """
+    log_file = pathlib.Path(conf_folder) / LOG_FILE
+    if log_file.exists():
+        if force:
+            log_file.unlink()
+        else:
+            raise FileExistsError(
+                f"Log file already exists: {log_file} evolven3fit_new has already been run?"
+            )
+    log_file = logging.FileHandler(log_file)
+    stdout_log = logging.StreamHandler(sys.stdout)
+    for log in [log_file, stdout_log]:
+        log.setLevel(LOGGING_SETTINGS["level"])
+        log.setFormatter(
+            logging.Formatter(LOGGING_SETTINGS["formatter"])
+        )
+    for logger in (log, *[logging.getLogger("eko")]):
+        logger.handlers = []
+        logger.setLevel(LOGGING_SETTINGS["level"])
+        logger.addHandler(log_file)
+        logger.addHandler(stdout_log)
+
+    usr_path = pathlib.Path(conf_folder)
+    initial_PDFs_dict = load_fit(usr_path)
+    x_grid = np.array(
+        initial_PDFs_dict[list(initial_PDFs_dict.keys())[0]]["xgrid"]
+    ).astype(np.float)
+    theoryID = utils.get_theoryID_from_runcard(usr_path)
+    theory, op = eko_utils.construct_eko_cards(
+        theoryID, q_fin, q_points, x_grid, op_card_dict, t_card_dict
+    )
+    if eko_path is not None:
+        eko_path = pathlib.Path(eko_path)
+        log.info(f"Loading eko from : {eko_path}")
+        eko_op = output.Output.load_tar(eko_path)
+    else:
+        try:
+            log.info(f"Loading eko from theory {theoryID}")
+            theory_eko_path = (API.theoryid(theoryid = theoryID).path)/'eko.tar'
+            eko_op = output.Output.load_tar(theory_eko_path)
+        except:
```

Comment:
Here however we definitely don't want to catch every exception as even an error on the log will lead to computing the whole eko! We only want to compute it if it is really not found.

Reply:
I missed it, sorry...

Reply:
Right, thanks. Now it is `except FileNotFoundError:` which is what we want, right?

Reply:
Probably yes. Note that you can even "force it" and instead of having a try-except do

```
if theory_eko_path.exists():
    load_tar()
else:
    calculate_eko()
```

Reply:
`try... except` is considered more Pythonic